---
title: "Create mater table for WTC-11 dataset"
author: "Vladimir Souza"
date: "5/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



# Introduction

Create the master table for WTC-11 data.  

Compare: DeepVariant, Clair3, and NanoCaller combined with SNCR, flagCorrection or alone, and GATK with SNCR.  

Iso-Seq:  
  * only primary alignments (secondary and supplementary alignments were filtered out);  
  * duplicates not marked/removed.  

sort-read:  
  * VCF downloaded from Allen Institute;  
  * only chromosomes chr1, chr2, ..., chr22, chrX, chrY;  
  * intronic regions removed.  




# Input variables

```{r}
### methods to validate
# name of the methods to validate
METHOD_NAMES <- c("dv", "dv_s", "dv_s_fc",
                  "c3", "c3_s", "c3_s_fc", "c3_mix",
                  "gatk_s",
                  "nc_s")
# name of the dataset used with the methods to validate
METHOD_DATASET_NAME <- "isoSeq"
# VCF files
METHOD_VCF_FILES <- c(
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/deepvariant/dv/deepvariant_calls_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/deepvariant/dv_sncr/deepvariant_calls_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/deepvariant/dv_sncr_fc/deepvariant_calls_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/clair3/c3/pileup_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/clair3/c3_sncr/pileup_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/clair3/c3_sncr_fc/pileup_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/clair3/mix/pileup_pass_mix_nodup.recode.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/gatk/isoSeq_wtc11.recal_pass.vcf.gz",
  "/home/vbarbo/project_2021/paper_analysis/wtc11/variant_calling_from_isoseq/nanocaller/nc_sncr/variant_calls.final.vcf.gz"
)
# BAM of the data
METHOD_BAM_FILE <- "/home/vbarbo/project_2021/paper_analysis/wtc11/data_manipulation/aln_s.bam"

### ground-truth
# name
TRUTH_NAME <- "allen"
# name of the dataset used to generate the ground-truth
TRUTH_DATASET_NAME <- "shortRead"
# VCF file
TRUTH_VCF_FILE <- "/home/vbarbo/project_2021/paper_analysis/wtc11/ground_truth/3546dc62_AH77TTBBXX_DS-229105_GCCAAT_recalibrated_subsetChromosomes_pass.vcf.gz"
# BAM file
TRUTH_BAM_FILE <- "/home/vbarbo/project_2021/paper_analysis/wtc11/ground_truth/84773251_trimmed.AH77TTBBXX_DS-229105_GCCAAT.sorted.rg.final.1.bam"


### variables
MAX_DIST_FROM_SPLICE_SITE <- 20
THREADS <- 30

GENOME_REF_FILE <- "/home/vbarbo/project_2021/paper_analysis/reference/genome/GRCh38.p13_all_chr.fasta"
```




# libraries

```{r}
library(lrRNAseqBenchmark)
library(GenomicAlignments)
library(dplyr)
library(sarlacc)
```




# Create the master table

## Initiate the master table

Get variant positions and which method could call them.

```{r}
dat <- initiate_master_table(
  METHOD_VCF_FILES[1],
  METHOD_VCF_FILES[2],
  METHOD_VCF_FILES[3],
  METHOD_VCF_FILES[4],
  METHOD_VCF_FILES[5],
  METHOD_VCF_FILES[6],
  METHOD_VCF_FILES[7],
  METHOD_VCF_FILES[8],
  METHOD_VCF_FILES[9],
  TRUTH_VCF_FILE,
  method_names=c(METHOD_NAMES, TRUTH_NAME)
)
```


## Add QD tag from ground-truth VCF

```{r}
# dat <- add_info_tag_from_vcf(dat, "qd_allen", "QD", TRUTH_VCF_FILE)
```


## Add dp tag for the ground truth.

```{r}
# dat <- add_info_tag_from_vcf(dat, "dp_info_allen", "DP", TRUTH_VCF_FILE)
```


## Add QUAL tag from VCF

```{r}
dat <- add_qual_from_vcf(dat, "dv", METHOD_VCF_FILES[1])
dat <- add_qual_from_vcf(dat, "dvS", METHOD_VCF_FILES[2])
dat <- add_qual_from_vcf(dat, "dvSFc", METHOD_VCF_FILES[3])
dat <- add_qual_from_vcf(dat, "c3", METHOD_VCF_FILES[4])
dat <- add_qual_from_vcf(dat, "c3S", METHOD_VCF_FILES[5])
dat <- add_qual_from_vcf(dat, "c3SFc", METHOD_VCF_FILES[6])
dat <- add_qual_from_vcf(dat, "c3Mix", METHOD_VCF_FILES[7])
dat <- add_qual_from_vcf(dat, "gatkS", METHOD_VCF_FILES[8])
dat <- add_qual_from_vcf(dat, "ncS", METHOD_VCF_FILES[9])
```


## get splice site positions from the BAM file and add columns about splice sites

```{r}
method_bam <- readGAlignments(METHOD_BAM_FILE)
splice_sites <- get_splice_sites_info(method_bam, THREADS)

# save `splice_sites` to file
save(splice_sites, file="/home/vbarbo/project_2021/paper_analysis/extra_files/splice_sites_isoseq_wtc11.RData")

dat <- add_splice_site_info_to_master_table(
  dat,
  splice_sites,
  MAX_DIST_FROM_SPLICE_SITE
)
```


## Add the read coverage (from the BAM file) of each variant

Add iso-seq read coverage.

```{r}
method_coverage <- coverage(METHOD_BAM_FILE)

# save `method_coverage` to file
# since this object takes a long time to be generated, next time load it instead
save(method_coverage, file="/home/vbarbo/project_2021/paper_analysis/extra_files/method_coverage_wtc11.RData")

dat <- add_read_coverage_from_bam_to_master_table(
  dat,
  method_coverage,
  METHOD_DATASET_NAME
)
```

Add short-read coverage (ground truth).

```{r}
### this takes a long time to run.
### we need to genereate the coverage of the ground-truth bam. since it is so big to store in a single object, we
### split it into two ones. a for loop is used to get the coverage per chromosome. The object `truth_coverage_1`
### stores the coverage of chromosomes 1--12, and `truth_coverage_2` of 13--22 + chrX + chrY.
### some reading on how to read bam by chromosomes: https://www.biostars.org/p/391925/

get_coverage_of_specified_chrms <- function(seqnames, obj_name){
  times <- NULL
  env[[obj_name]] <- RleList()
  
  for(seqname in seqnames){
    TIME <- Sys.time()
    
    i <- which(seqnames == seqname)
    
    end_range <- seqlengths( bamInfo[seqname] )
    param <- ScanBamParam(what=c("pos", "qwidth"),
                          which=GRanges(seqname, IRanges(1, end_range)),
                          flag=scanBamFlag(isUnmappedQuery=FALSE))
    x <- coverage( readGAlignments(TRUTH_BAM_FILE, param=param) )
    x <- x[ names(x) == seqname ]
    env[[obj_name]] <- c(env[[obj_name]], x)
    
    times[i] <- Sys.time() - TIME
    names(times)[i] <- seqname
    print(times)
  }
}

bamInfo <- seqinfo(BamFile(TRUTH_BAM_FILE))
env <- environment()

chromoso_names <- paste0("chr", 1:12)
get_coverage_of_specified_chrms(chromoso_names, "truth_coverage_1")

chromoso_names <- c( paste0("chr", 13:22), "chrX", "chrY" )
get_coverage_of_specified_chrms(chromoso_names, "truth_coverage_2")

# save `truth_coverage_1` and `truth_coverage_2` objects to a RData file
# since these objects take a long time to be generated, next time load them instead
save(truth_coverage_1, truth_coverage_2, file="/home/vbarbo/project_2021/paper_analysis/extra_files/truth_coverage_1_2_wtc11.RData")

### add short-read coverage to master table. must do it in two steps: first the chromosomes in truth_coverage_1 and lately the ones in truth_coverage_2
dat_1 <- filter(dat, chrm %in% names(truth_coverage_1))
dat_2 <- filter(dat, chrm %in% names(truth_coverage_2))

dat_1 <- add_read_coverage_from_bam_to_master_table(
  dat_1,
  truth_coverage_1,
  TRUTH_DATASET_NAME
)

dat_2 <- add_read_coverage_from_bam_to_master_table(
  dat_2,
  truth_coverage_2,
  TRUTH_DATASET_NAME
)

dat <- rbind(dat_1, dat_2) %>% 
  arrange(chrm, pos)
```


## Add the number of N-cigar reads per site

```{r}
dat <- add_number_of_n_cigar_reads_to_master_table(
  dat, 
  method_bam, 
  METHOD_DATASET_NAME
)
```


## Add column to classify the method calls (compare to the ground-truth)

```{r}
### method 1
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[1],
  TRUTH_NAME
)
### method 2
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[2],
  TRUTH_NAME
)
### method 3
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[3],
  TRUTH_NAME
)
### method 4
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[4],
  TRUTH_NAME
)
### method 5
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[5],
  TRUTH_NAME
)
### method 6
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[6],
  TRUTH_NAME
)
### method 7
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[7],
  TRUTH_NAME
)
### method 8
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[8],
  TRUTH_NAME
)
### method 9
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[9],
  TRUTH_NAME
)
```


## Add columns to inform the variant type according each method and the ground truth

For the ground-truth

```{r}
### ground truth
dat <- add_variant_type_to_master_table(dat, TRUTH_VCF_FILE, TRUTH_NAME)

### methods
dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[1],
                                        METHOD_NAMES[1])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[2],
                                        METHOD_NAMES[2])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[3],
                                        METHOD_NAMES[3])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[4],
                                        METHOD_NAMES[4])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[5],
                                        METHOD_NAMES[5])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[6],
                                        METHOD_NAMES[6])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[7],
                                        METHOD_NAMES[7])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[8],
                                        METHOD_NAMES[8])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[9],
                                        METHOD_NAMES[9])
```


## Add columns to inform whether the variant is a indel or not (-1, 0, 1, or NA)

* 1 = is indel
* 0 = is not indel
* -1 = non defined (heterozygous alternative) (truth's gt has priority over the method's)
* NA = variant not calls by the method or the truth

```{r}
### truth
dat <- is_indel_method(dat, TRUTH_NAME)

### dv
dat <- is_indel_method(dat, TRUTH_NAME, "dv")

### dv_s
dat <- is_indel_method(dat, TRUTH_NAME, "dv_s")

### dv_s_fc
dat <- is_indel_method(dat, TRUTH_NAME, "dv_s_fc")

### c3
dat <- is_indel_method(dat, TRUTH_NAME, "c3")

### c3_s
dat <- is_indel_method(dat, TRUTH_NAME, "c3_s")

### c3_s_fc
dat <- is_indel_method(dat, TRUTH_NAME, "c3_s_fc")

### c3_mix
dat <- is_indel_method(dat, TRUTH_NAME, "c3_mix")

### gatk_s
dat <- is_indel_method(dat, TRUTH_NAME, "gatk_s")

### nc_s
dat <- is_indel_method(dat, TRUTH_NAME, "nc_s")
```


## Add variant/sequencing_error dencity per variant

```{r}
### TRUTH_NAME
dat <- add_variant_density_of_a_method(dat, 201, TRUTH_NAME)

### all methods
dat <- add_variant_density_of_a_method(dat, 201, METHOD_NAMES)
```


## add genotype

```{r}
dat <- add_format_tag_from_vcf(dat, TRUTH_NAME, "GT", TRUTH_VCF_FILE)
dat <- add_format_tag_from_vcf(dat, "dv_s_fc", "GT", METHOD_VCF_FILES[3])
dat <- add_format_tag_from_vcf(dat, "c3_mix", "GT", METHOD_VCF_FILES[7])
dat <- add_format_tag_from_vcf(dat, "gatk_s", "GT", METHOD_VCF_FILES[8])
dat <- add_format_tag_from_vcf(dat, "nc_s", "GT", METHOD_VCF_FILES[9])

sel_cols <- paste0( "gt_", c(TRUTH_NAME, METHOD_NAMES[c(3,7,8,9)]) )
k <- is.na( dat[,sel_cols] )
dat[,sel_cols] [k] <- "0/0"
```


## add homopolimer length (from reference FASTA), but only for 0/1 and 1/1 genotypes and indels

```{r}
genome_ref <- readDNAStringSet(GENOME_REF_FILE)
homopolymers <- homopolymerFinder(genome_ref)
names(homopolymers) <- sub("(^chr[0-9]+|X|Y).*", "\\1", names(homopolymers))

# save `homopolymers` object to a RData file
# since this object takes a long time to be generated, next time load it instead
save(homopolymers, file="/home/vbarbo/project_2021/paper_analysis/extra_files/homopolymers_GRCh38.p13_all_chr.RData")

dat <- add_homopolymer_length_when_indels(dat, homopolymers)
```


## Take a look at the master table

```{r}
head(dat)
dat_bk <- dat
```


## Save master table to a file

```{r}
mt_wtc11_allMethods_notFiltered <- dat
save(mt_wtc11_allMethods_notFiltered,
     file="/home/vbarbo/project_2021/paper_analysis/extra_files/master_tables/mt_wtc11_allMethods_notFiltered_v7.RData",
     compress="xz")
```




# Filtering master table

Sites to keep:  
* short-read variant density: x <= 3                          (ignore positions)  
* ignore intronic regions (isoSeq_coverage > 0)               (ignore positions)  
* DeepVariant QUAL: x >= 15                                   (remove method calling)  
* short-read coverare : quantil 5% <= x <= 95%                (ignore positions)  

```{r}
# no filtering
dat1 <- dat
dim(dat1)

shortread_cover_quantiles <- quantile(dat1$shortRead_coverage, probs=c(.05, .95))

# variant/sequence_error density -- filter using the ground truth
k <- paste0("variantDensity_", TRUTH_NAME)
k <- !( dat1[,k] > 3 )
dat1 <- dat1[k,]
dim(dat1)

# ignore intronic regions (iso-seq coverage equal to zero)
dat1 <- filter(dat1, isoSeq_coverage>0)
dim(dat1)

# ignore regions based on short-read coverage quantiles: keep 5% <= x <= 95%
dat1 <- filter(dat1, shortRead_coverage >= shortread_cover_quantiles[1] &
                 shortRead_coverage <= shortread_cover_quantiles[2])
dim(dat1)

# Filter out DeepVariant calls by QUAL -- but keep the position
# dv
k <- which( dat1$qual_dv < 15 )
dat1$in_dv[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[1],
  TRUTH_NAME,
  replace_column=TRUE
)
# dv_s
k <- which( dat1$qual_dvS < 15 )
dat1$in_dv_s[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[2],
  TRUTH_NAME,
  replace_column=TRUE
)
# dv_s_fc
k <- which( dat1$qual_dvSFc < 15 )
dat1$in_dv_s_fc[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[3],
  TRUTH_NAME,
  replace_column=TRUE
)


### don't filter by Clair3 QUAL

### don't filter by NanoCaller QUAL
```

save filtered master table

```{r}
mt_wtc11_allMethods_filtered <- dat1
save(mt_wtc11_allMethods_filtered,
     file="/home/vbarbo/project_2021/paper_analysis/extra_files/master_tables/mt_wtc11_allMethods_filtered_v7.RData",
     compress="xz")
```


