#!/usr/bin/env Rscript


# This script is to fix problematic CIGAR strings of BAM files
# generated by GATK's SplitNCigarReads function. That is, when there 
# are repeated "S" (soft clipping) in a row, merge them. For example, 
# "1000M20S30S" becomes "1000M50S".
# 
# This function may make BAM files suitable for Picard's 
# AddOrReplaceReadGroups
# 
# The input parameters are (following the right order):
# * `INPUT_BAM`, path of the input BAM file;
# * `OUTPUT_BAM`, path of the output BAM file;
# * `THREADS`, number of cores.
#
# To run fixSoftClippingOfBAM:
# Rscript $PATH_TO_REPO/fixSoftClippingOfBAM.r \
#   $INPUT_BAM \
#   $OUTPUT_BAM \
#   $THREADS




### input variables
INPUT_BAM=commandArgs(TRUE)[1]
OUTPUT_BAM=commandArgs(TRUE)[2]
THREADS=as.integer(commandArgs(TRUE)[3])



### time when of start
time1 <- Sys.time()


### test inputs
if( !file.exists(INPUT_BAM) ) stop( gettextf("File %s doesn't exist", INPUT_BAM) )
output_dir <- dirname(OUTPUT_BAM)
if( !dir.exists(output_dir) ) stop( gettextf("Directory %s doesn't exist", output_dir) )


### load packages
suppressPackageStartupMessages( library(foreach) )
suppressPackageStartupMessages( library(doParallel) )
suppressPackageStartupMessages( library(GenomicAlignments) )


### create root temp dir
temp_dir <- tempfile(pattern="temp_dir_", tmpdir=output_dir)
dir.create(temp_dir)


### write sam header from bam
sam_header <- tempfile(pattern="sam_header_", tmpdir=temp_dir, fileext=".sam")
cmd <- gettextf("samtools view -H %s > %s", INPUT_BAM, sam_header)
system(cmd)


### get chromosome names
cmd <- gettextf("perl -lne '/SN:(\\S+)/ and print $1' %s", sam_header)
chr_names <- system(cmd, intern=TRUE)


### define some variables
# temp dir for each chromosome
chr_temp_dirs <- file.path(temp_dir, chr_names)
# faulty sam body files
faulty_sam_files <- tempfile(pattern="faulty_sam_file_", tmpdir=chr_temp_dirs, fileext=".sam")
# fixed qnames txt files
fixed_cigar_files <- tempfile(pattern="fixed_cigar_", tmpdir=chr_temp_dirs, fileext=".txt")
# fixed sam body files
fixed_sam_files <- tempfile(pattern="fixed_sam_files_", tmpdir=chr_temp_dirs, fileext=".sam")
# headered fixed sam files
headered_fixed_sam_files <- tempfile(pattern="headered_fixed_sam_files_", tmpdir=chr_temp_dirs, fileext=".sam")
# fixed bam files
bam_files <- gettextf("%s/%s.bam", temp_dir, chr_names)



### make fixed-qnames sam files for each chromosome
make_qnames_unique <- function(chr_temp_dirs_i, chr_names_i, faulty_sam_files_i,
                               fixed_cigar_files_i, fixed_sam_files_i,
                               headered_fixed_sam_files_i, bam_files_i,
                               sam_header){
  
  ### create a temp dir for the chromosome
  dir.create(chr_temp_dirs_i)
  
  
  ### create a SAM file for each chromosome
  cmd <- gettextf("samtools view %s %s > %s", INPUT_BAM, chr_names_i, faulty_sam_files_i)
  system(cmd)
  
  
  ### load cigar
  cmd <- gettextf("cut -f 6 %s", faulty_sam_files_i)
  cig <- system(cmd, intern=TRUE)
  
  
  ### find repeated "S" in a row
  cigLetters <- explodeCigarOps(cig)
  cigLettersPaste <- sapply(cigLetters, paste, collapse="")
  to_fix <- grep("SS", cigLettersPaste)
  
  if( length(to_fix)==0 ){
    ### if SAM is all right, just add the headers to it
    cmd <- gettextf("cat %s %s > %s",
                    sam_header, faulty_sam_files_i, headered_fixed_sam_files_i)
    system(cmd)
    
  }else{
    ### fix cigar strings
    cigLengths <- explodeCigarOpLengths(cig[to_fix])
    pos_to_fix <- gregexpr("(S)\\1{1,}", cigLettersPaste[to_fix])
    fixed_cig <- mapply(function(let, len, ptf){
      pos_extended <- NULL
      for(i in seq_along(ptf)){
        ptf_i <- as.integer(ptf)[i]
        ptfLen_i <- attr(ptf, "match.length")[i]
        k <- ptf_i:(ptf_i + ptfLen_i-1)
        len[ k[1] ] <- sum(len[k])
        pos_extended <- c(pos_extended, k[-1])
      }
      let <- let[-pos_extended]
      len <- len[-pos_extended]
      paste( c( matrix( c(len, let), nrow=2, byrow=TRUE ) ), collapse="" )
    }, cigLetters[to_fix], cigLengths, pos_to_fix)
    
    
    ### replace the faulty for the fixed cigar strings and save to a file
    cig[to_fix] <- fixed_cig
    write.table(cig, file=fixed_cigar_files_i, quote=FALSE, row.names=FALSE, col.names=FALSE)
    
    
    ### fix SAM file
    cmd <- gettextf("awk 'FNR==NR{a[NR]=$1;next}{$6=a[FNR]}1' OFS='\t' %s %s > %s",
                    fixed_cigar_files_i, faulty_sam_files_i, fixed_sam_files_i)
    system(cmd)
    
    
    ### add headers to SAM
    cmd <- gettextf("cat %s %s > %s",
                    sam_header, fixed_sam_files_i, headered_fixed_sam_files_i)
    system(cmd)
  }
  
  
  ### convert sam to bam
  cmd <- gettextf("samtools view -S -b %s > %s",
                  headered_fixed_sam_files_i, bam_files_i)
  system(cmd)
  
  
  ### delete chromosome temp files and mark dir as done
  unlink(chr_temp_dirs_i, recursive=TRUE)
}



### the maximun number of cores is the number of chomosomes
threads <- min( THREADS, length(chr_names) )


### run function using multiple cores
registerDoParallel(cores=threads)
invisible(
  foreach(chr_temp_dirs_i=chr_temp_dirs,
          chr_names_i=chr_names,
          faulty_sam_files_i=faulty_sam_files,
          fixed_cigar_files_i=fixed_cigar_files,
          fixed_sam_files_i=fixed_sam_files,
          headered_fixed_sam_files_i=headered_fixed_sam_files,
          bam_files_i=bam_files) %dopar%
    make_qnames_unique(chr_temp_dirs_i,
                       chr_names_i,
                       faulty_sam_files_i,
                       fixed_cigar_files_i,
                       fixed_sam_files_i,
                       headered_fixed_sam_files_i,
                       bam_files_i,
                       sam_header)
)



### if `OUTPUT_BAM` doesn't contain the bam extension, add it
if( ! grepl("\\.bam$", OUTPUT_BAM) )
  OUTPUT_BAM <- paste0(OUTPUT_BAM, ".bam")


### merge bams
all_bams_to_merge <- paste(bam_files, collapse=" ")
cmd <- gettextf("samtools merge -c -p -f -@ %i %s %s", THREADS, OUTPUT_BAM, all_bams_to_merge)
system(cmd)


### delete temp dir
unlink(temp_dir, recursive=TRUE)


### report time spent
time2 <- Sys.time()
Time <- time2 - time1
Time <- round( as.numeric(Time, units = "mins"), 2 )
cat( gettextf("\nmakeQnamesUnique finished after %s minutes.\n\n", Time) )


